presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/MergeJoinOptimizer.java

```diff
@@ -0,0 +1,128 @@
+import com.facebook.presto.sql.planner.PlanVariableAllocator;
+import com.facebook.presto.sql.planner.TypeProvider;
+import com.facebook.presto.sql.planner.plan.JoinNode;
+import com.facebook.presto.sql.planner.plan.MergeJoinNode;
+import com.facebook.presto.sql.planner.plan.SimplePlanRewriter;
+
+import java.util.List;
+
+import static com.facebook.presto.SystemSessionProperties.preferMergeJoin;
+import static com.facebook.presto.common.block.SortOrder.ASC_NULLS_FIRST;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static java.util.Objects.requireNonNull;
+
+public class MergeJoinOptimizer
+        implements PlanOptimizer
+{
+    private final Metadata metadata;
+    private final SqlParser parser;
+
+    public MergeJoinOptimizer(Metadata metadata, SqlParser parser)
+    {
+        this.metadata = requireNonNull(metadata, "metadata is null");
+        this.parser = requireNonNull(parser, "parser is null");
+    }
+
+    @Override
+    public PlanNode optimize(PlanNode plan, Session session, TypeProvider type, PlanVariableAllocator variableAllocator, PlanNodeIdAllocator idAllocator, WarningCollector warningCollector)
+    {
+        requireNonNull(plan, "plan is null");
+        requireNonNull(session, "session is null");
+        requireNonNull(variableAllocator, "variableAllocator is null");
+        requireNonNull(idAllocator, "idAllocator is null");
+
+        if (preferMergeJoin(session)) {
+            return SimplePlanRewriter.rewriteWith(new MergeJoinOptimizer.Rewriter(variableAllocator, idAllocator, metadata, session), plan, null);
+        }
+        return plan;
+    }
+
+    private class Rewriter
+            extends SimplePlanRewriter<Void>
+    {
+        private final PlanNodeIdAllocator idAllocator;
+        private final Metadata metadata;
+        private final Session session;
+        private final TypeProvider types;
+
+        private Rewriter(PlanVariableAllocator variableAllocator, PlanNodeIdAllocator idAllocator, Metadata metadata, Session session)
+        {
+            this.idAllocator = requireNonNull(idAllocator, "idAllocator is null");
+            this.metadata = requireNonNull(metadata, "metadata is null");
+            this.session = requireNonNull(session, "session is null");
+            this.types = variableAllocator.getTypes();
+        }
+
+        @Override
+        public PlanNode visitJoin(JoinNode node, RewriteContext<Void> context)
+        {
+            // When we have a plan that looks like:
+            // JoinNode
+            //- TableScanA
+            //- TableScanB
+
+            // We check the data properties of TableScanA and TableScanB to see if they meet requirements for merge join:
+            // 1. If so, we replace the JoinNode to MergeJoinNode
+            // MergeJoinNode
+            //- TableScanA
+            //- TableScanB
+
+            // 2. If not, we don't optimize
+
+            if (node.getLeft() instanceof TableScanNode
```

Comment:
Merge join doesn't require both sides to be table scans.
For example, we can have
MergeJoinNode
- StreamingAggregation
  - TableScan
- Sort
  - TableScan 

There could be projects/StreamingAggregation/Sort or another merge join above the TableScan.
As long as the child subplan can guarantee the sorted property, we should be able to enable merge join.